import sys
from PyQt6 import QtWidgets, uic, QtCore, QtGui 
from google import genai
from google.genai.errors import APIError
from google.genai import types
import os
# playsound, gTTS, ì„ì‹œ íŒŒì¼ ì²˜ë¦¬ë¥¼ ìœ„í•œ ëª¨ë“ˆì„ ê°€ì ¸ì˜µë‹ˆë‹¤.
from gtts import gTTS
from playsound import playsound
import uuid # ê³ ìœ í•œ ì„ì‹œ íŒŒì¼ ì´ë¦„ì„ ìœ„í•´ í•„ìš”

# =================================================================
# 1. ì„¤ì • ë° ì´ˆê¸°í™”
# =================================================================

# 1. UI íŒŒì¼ ì´ë¦„ ì„¤ì •
UI_FILE_NAME = "Mygemini.ui"

# ğŸŒŸ ì¤‘ìš”: API í‚¤ë¥¼ ì—¬ê¸°ì— ì…ë ¥í•˜ì„¸ìš”. ğŸŒŸ
GEMINI_API_KEY = "GEMINI_API_KEY"

# 2. Gemini í´ë¼ì´ì–¸íŠ¸ ì´ˆê¸°í™”
if not GEMINI_API_KEY:
    print("ì¹˜ëª…ì  ì˜¤ë¥˜: GEMINI_API_KEYê°€ ì„¤ì •ë˜ì§€ ì•Šì•˜ìŠµë‹ˆë‹¤.")
    sys.exit(1)
    
try:
    client = genai.Client(api_key=GEMINI_API_KEY)
    MODEL_NAME = "gemini-2.5-flash"
except ValueError as e:
    print(f"í´ë¼ì´ì–¸íŠ¸ ì´ˆê¸°í™” ì˜¤ë¥˜: API í‚¤ê°€ ì˜ëª»ë˜ì—ˆê±°ë‚˜ ëˆ„ë½ë˜ì—ˆìŠµë‹ˆë‹¤. {e}")
    sys.exit(1)


# =================================================================
# 2. Gemini API í˜¸ì¶œì„ ìœ„í•œ ì›Œì»¤ ìŠ¤ë ˆë“œ (QThread)
# =================================================================

class GeminiWorker(QtCore.QThread):
    response_ready = QtCore.pyqtSignal(str, str) 
    error_occurred = QtCore.pyqtSignal(str, str) 

    def __init__(self, client, model_name, chat_history, user_question):
        super().__init__()
        self.client = client
        self.model_name = model_name
        self.chat_history = chat_history 
        self.user_question = user_question
        
        self.config = types.GenerateContentConfig(
            system_instruction="You are a helpful assistant. Please answer all questions in Korean."
        )

    def run(self):
        try:
            response = self.client.models.generate_content(
                model=self.model_name,
                contents=self.chat_history,
                config=self.config
            )
            
            gemini_response = response.text
            
            self.chat_history.append({"role": "model", "parts": [{"text": gemini_response}]})

            self.response_ready.emit(self.user_question, gemini_response)

        except APIError as e:
            self.error_occurred.emit("API ì˜¤ë¥˜", f"Gemini ì„œë²„ì— ì—°ê²°í•  ìˆ˜ ì—†ìŠµë‹ˆë‹¤. (ì˜¤ë¥˜: {e})")
        except Exception as e:
            self.error_occurred.emit("ì˜ˆê¸°ì¹˜ ì•Šì€ ì˜¤ë¥˜", str(e))


# =================================================================
# 3. ë©”ì¸ ì• í”Œë¦¬ì¼€ì´ì…˜ í´ë˜ìŠ¤
# =================================================================

class GeminiApp(QtWidgets.QDialog):
    def __init__(self):
        super().__init__()
        
        try:
            uic.loadUi(UI_FILE_NAME, self)
        except FileNotFoundError:
            print(f"ì˜¤ë¥˜: {UI_FILE_NAME} íŒŒì¼ì„ ì°¾ì„ ìˆ˜ ì—†ìŠµë‹ˆë‹¤. ê²½ë¡œë¥¼ í™•ì¸í•´ ì£¼ì„¸ìš”.")
            sys.exit(1)
            
        self.setWindowTitle(f"Gemini Q&A - ëª¨ë¸: {MODEL_NAME}")
        
        # ìœ„ì ¯ ì°¾ê¸° (ì´ì „ ì½”ë“œì™€ ë™ì¼)
        self.btnSend = self.findChild(QtWidgets.QPushButton, 'btnSend')
        self.lineEditMyQuestion = self.findChild(QtWidgets.QLineEdit, 'lineEditMyQuestion')
        
        self.lblAnswer = self.findChild(QtWidgets.QTextEdit, 'lblAnswer') 
        if not self.lblAnswer:
            self.lblAnswer = self.findChild(QtWidgets.QPlainTextEdit, 'lblAnswer')
        
        if not self.btnSend or not self.lblAnswer or not self.lineEditMyQuestion:
            print("ì¹˜ëª…ì  ì˜¤ë¥˜: UI íŒŒì¼ì—ì„œ í•„ìˆ˜ ìœ„ì ¯ì„ ì°¾ì„ ìˆ˜ ì—†ìŠµë‹ˆë‹¤.")
            sys.exit(1)

        # ìœ„ì ¯ ì´ë²¤íŠ¸ ì—°ê²° ë° ì„¤ì •
        self.btnSend.clicked.connect(self.generate_response)
        self.lblAnswer.setReadOnly(True)

        self.chat_history = []
        self.gemini_worker = None 
        
        self.lblAnswer.setText("[Mygemini] ë¬´ì—‡ì„ ë„ì™€ë“œë¦´ê¹Œìš”?")

        self.show()

    def generate_response(self):
        user_question = self.lineEditMyQuestion.text().strip()
        
        if not user_question:
            return
        
        user_message = f"[ì§ˆë¬¸] {user_question}\n"
        self.lblAnswer.append(user_message)
        
        loading_message = "[Mygemini] ì‘ë‹µì„ ìƒì„±í•˜ëŠ” ì¤‘ì…ë‹ˆë‹¤..."
        self.lblAnswer.append(loading_message)
        QtWidgets.QApplication.processEvents() 
        
        self.chat_history.append({"role": "user", "parts": [{"text": user_question}]})

        # Gemini API í˜¸ì¶œ (QThread ì‚¬ìš©)
        self.gemini_worker = GeminiWorker(
            client=client,
            model_name=MODEL_NAME,
            chat_history=self.chat_history,
            user_question=user_question
        )
        
        self.gemini_worker.response_ready.connect(self.handle_response)
        self.gemini_worker.error_occurred.connect(self.handle_error)
        
        self.gemini_worker.start()
        
        self.lineEditMyQuestion.clear()
        
    def handle_response(self, user_question, gemini_response):
        """API ì‘ë‹µì„ ë°›ì•„ UIì— ìµœì¢…ì ìœ¼ë¡œ í‘œì‹œí•˜ê³  ë¡œë”© ë©”ì‹œì§€ë¥¼ ì œê±°í•˜ë©° ìŒì„± ì¶œë ¥í•©ë‹ˆë‹¤."""
        
        formatted_output = f"[Mygemini] {gemini_response}\n"
        
        # ë¡œë”© ë©”ì‹œì§€ë¥¼ ì œê±°í•©ë‹ˆë‹¤.
        cursor = self.lblAnswer.textCursor()
        cursor.movePosition(QtGui.QTextCursor.MoveOperation.End)
        cursor.movePosition(QtGui.QTextCursor.MoveOperation.StartOfLine, QtGui.QTextCursor.MoveMode.KeepAnchor)
        cursor.removeSelectedText()
        
        # ìµœì¢… ì‘ë‹µì„ ì¶”ê°€í•©ë‹ˆë‹¤.
        self.lblAnswer.append(formatted_output)
        
        # ğŸŒŸ ì‘ë‹µì„ ìŒì„±ìœ¼ë¡œ ì¶œë ¥í•©ë‹ˆë‹¤.
        self.text_to_speech(gemini_response)

        # ìŠ¤ë ˆë“œ ì¢…ë£Œ ë° ì •ë¦¬
        self.gemini_worker.quit()
        self.gemini_worker.wait()
        
    def handle_error(self, error_type, error_message):
        """API ì˜¤ë¥˜ ë°œìƒ ì‹œ UIì— ì˜¤ë¥˜ ë©”ì‹œì§€ë¥¼ í‘œì‹œí•˜ê³  ë¡œë”© ë©”ì‹œì§€ë¥¼ ì œê±°í•©ë‹ˆë‹¤."""
        
        if self.chat_history and self.chat_history[-1].get("role") == "user":
             self.chat_history.pop()

        formatted_output = f"[Mygemini] {error_type} ë°œìƒ: {error_message}\n"
        
        # ë¡œë”© ë©”ì‹œì§€ë¥¼ ì œê±°í•˜ê³  ì˜¤ë¥˜ ë©”ì‹œì§€ë¥¼ ì¶”ê°€í•©ë‹ˆë‹¤.
        cursor = self.lblAnswer.textCursor()
        cursor.movePosition(QtGui.QTextCursor.MoveOperation.End)
        cursor.movePosition(QtGui.QTextCursor.MoveOperation.StartOfLine, QtGui.QTextCursor.MoveMode.KeepAnchor)
        cursor.removeSelectedText()
        
        self.lblAnswer.append(formatted_output)
        
        # ğŸŒŸ ì˜¤ë¥˜ ë©”ì‹œì§€ë¥¼ ìŒì„±ìœ¼ë¡œ ì¶œë ¥í•©ë‹ˆë‹¤.
        self.text_to_speech(f"{error_type} ë°œìƒ: {error_message}")


        # ìŠ¤ë ˆë“œ ì¢…ë£Œ ë° ì •ë¦¬
        self.gemini_worker.quit()
        self.gemini_worker.wait()

    # =================================================================
    # ğŸŒŸ ìˆ˜ì •ëœ ìŒì„± ì¶œë ¥ ë©”ì„œë“œ (gTTS + playsound ì‚¬ìš©, ì„ì‹œ íŒŒì¼ ì‚¬ìš©)
    # =================================================================
    def text_to_speech(self, text):
        """gTTSë¥¼ ì‚¬ìš©í•˜ì—¬ í…ìŠ¤íŠ¸ë¥¼ ìŒì„±ìœ¼ë¡œ ë³€í™˜í•˜ê³  playsoundë¡œ ì¬ìƒí•©ë‹ˆë‹¤."""
        
        try:
            print("ìŒì„± ìƒì„± ë° ì¬ìƒ ì‹œì‘...")
            # ì„ì‹œ íŒŒì¼ ì´ë¦„ ìƒì„± (ì¤‘ë³µ ë°©ì§€)
            filename = f"temp_speech_{uuid.uuid4().hex}.mp3"
            
            # 1. gTTS ê°ì²´ ìƒì„± (í•œêµ­ì–´ 'ko' ì„¤ì •) ë° íŒŒì¼ë¡œ ì €ì¥
            tts = gTTS(text=text, lang='ko')
            tts.save(filename)
            
            # 2. playsoundë¡œ ì¬ìƒ (ì´ ë¶€ë¶„ì´ ë¸”ë¡œí‚¹ë©ë‹ˆë‹¤)
            playsound(filename)
            
            # 3. ì¬ìƒ í›„ íŒŒì¼ ì‚­ì œ (í´ë¦°ì—…)
            os.remove(filename)
            print("ìŒì„± ì¬ìƒ ì™„ë£Œ ë° íŒŒì¼ ì‚­ì œ.")
            
        except Exception as e:
            # ìŒì„± ì¶œë ¥ ì¤‘ ì˜¤ë¥˜ê°€ ë°œìƒí•´ë„ í”„ë¡œê·¸ë¨ì€ ê³„ì† ì‹¤í–‰ë˜ë„ë¡ ì˜ˆì™¸ ì²˜ë¦¬
            print(f"ìŒì„± ì¶œë ¥ ì˜¤ë¥˜ (gTTS/playsound): {e}")


# =================================================================
# 4. í”„ë¡œê·¸ë¨ ì‹¤í–‰
# =================================================================

if __name__ == '__main__':
    # ğŸŒŸ í”„ë¡œê·¸ë¨ ì‹œì‘ ì‹œ ë‚¨ì•„ìˆëŠ” temp_speech_*.mp3 íŒŒì¼ì„ ì‚­ì œí•©ë‹ˆë‹¤.
    for f in os.listdir('.'):
        if f.startswith('temp_speech_') and f.endswith('.mp3'):
            try:
                os.remove(f)
            except OSError as e:
                print(f"ì„ì‹œ íŒŒì¼ ì‚­ì œ ì˜¤ë¥˜ {f}: {e}")
                
    app = QtWidgets.QApplication(sys.argv)
    window = GeminiApp()

    sys.exit(app.exec())
